[GPU Process] [FormControls] Add ControlParts for SliderTrack and SliderThumb

shallawa opened the pull request:
#### 763b293216be504ad2cffa1429628f6ce5f08bc9
<pre>
[GPU Process] [FormControls] Add ControlParts for SliderTrack and SliderThumb
<a href="https://bugs.webkit.org/show_bug.cgi?id=249921">https://bugs.webkit.org/show_bug.cgi?id=249921</a>
rdar://103737847

Reviewed by Aditya Keerthi.

These ControlParts will handle drawing the SliderTrack and the SliderThumb form
controls (i.e. &lt;input type=&quot;range&quot; ...&gt;). For macOS, AppKit will be used to draw
the platform controls through SliderTrackMac and SliderThumbMac.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ControlStyle.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/controls/ControlStyle.h:
* Source/WebCore/platform/graphics/controls/SliderThumbPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
* Source/WebCore/platform/graphics/controls/SliderTrackPart.cpp: Added.
(WebCore::SliderTrackPart::create):
(WebCore::SliderTrackPart::SliderTrackPart):
(WebCore::SliderTrackPart::drawTicks const):
(WebCore::SliderTrackPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/SliderTrackPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
(WebCore::SliderTrackPart::thumbSize const):
(WebCore::SliderTrackPart::trackBounds const):
(WebCore::SliderTrackPart::tickRatios const):
(isType):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::sliderCell const):
(WebCore::ControlFactoryMac::createPlatformSliderThumb):
(WebCore::ControlFactoryMac::createPlatformSliderTrack):
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm: Added.
(WebCore::SliderThumbMac::SliderThumbMac):
(WebCore::SliderThumbMac::updateCellStates):
(WebCore::SliderThumbMac::rectForBounds const):
(WebCore::SliderThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
(WebCore::SliderTrackMac::owningSliderTrackPart const):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm: Added.
(WebCore::SliderTrackMac::SliderTrackMac):
(WebCore::SliderTrackMac::rectForBounds const):
(WebCore::trackGradientInterpolate):
(WebCore::SliderTrackMac::draw):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::createSliderTrackPartForRenderer):
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::extractControlStyleStatesForRenderer const):
(WebCore::RenderTheme::extractControlStyleForRenderer const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::TrackGradientInterpolate): Deleted.
(WebCore::RenderThemeMac::paintSliderTrack): Deleted.
(WebCore::RenderThemeMac::paintSliderThumb): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::encode):
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/258522@main">https://commits.webkit.org/258522@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b0d3b88793ba6274a691b8d02099a66f365417c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11340 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35265 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111519 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106179 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12320 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2254 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94563 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/109248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107976 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92709 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/94563 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24191 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94563 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/4884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25614 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5004 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2053 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/11050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45111 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5841 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6751 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
```diff
@@ -52,6 +52,22 @@
     return [[NSWorkspace sharedWorkspace] accessibilityDisplayShouldIncreaseContrast];
 }
 
+void ControlMac::updateEnabledState(NSCell* cell, const ControlStyle& style)
```
Nit: `NSCell *`.

shallawa replied:
Fixed.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm:
```diff
@@ -0,0 +1,84 @@
+    ASSERT(m_owningPart.type() == ControlPartType::SliderThumbHorizontal || m_owningPart.type() == ControlPartType::SliderThumbVertical);
+}
+
+void SliderThumbMac::updateCellStates(const FloatRect& rect, const ControlStyle& style)
```
I think this should call `ControlMac::updateCellStates` too.

shallawa replied:
You are right. Fixed.

pxlcoder commented on Source/WebCore/platform/graphics/controls/ControlStyle.h:
```diff
@@ -52,11 +52,13 @@ struct ControlStyle {
     unsigned fontSize { 12 };
     float zoomFactor { 1 };
     Color accentColor;
+    Color visitedDependentColor;
```
I would call this `textColor`, since it comes from `CSSPropertyColor`.

`visitedDependentColor` doesn't make sense, because there are a number of colors that can be dependent on the visited state.

shallawa replied:
The name was changed.

shallawa committed:
[GPU Process] [FormControls] Add ControlParts for SliderTrack and SliderThumb
https://bugs.webkit.org/show_bug.cgi?id=249921
rdar://103737847

Reviewed by Aditya Keerthi.

These ControlParts will handle drawing the SliderTrack and the SliderThumb form
controls (i.e. <input type="range" ...>). For macOS, AppKit will be used to draw
the platform controls through SliderTrackMac and SliderThumbMac.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ControlStyle.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/controls/ControlStyle.h:
* Source/WebCore/platform/graphics/controls/SliderThumbPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
* Source/WebCore/platform/graphics/controls/SliderTrackPart.cpp: Added.
(WebCore::SliderTrackPart::create):
(WebCore::SliderTrackPart::SliderTrackPart):
(WebCore::SliderTrackPart::drawTicks const):
(WebCore::SliderTrackPart::createPlatformControl):
* Source/WebCore/platform/graphics/controls/SliderTrackPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
(WebCore::SliderTrackPart::thumbSize const):
(WebCore::SliderTrackPart::trackBounds const):
(WebCore::SliderTrackPart::tickRatios const):
(isType):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::sliderCell const):
(WebCore::ControlFactoryMac::createPlatformSliderThumb):
(WebCore::ControlFactoryMac::createPlatformSliderTrack):
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
* Source/WebCore/platform/graphics/mac/controls/SliderThumbMac.mm: Added.
(WebCore::SliderThumbMac::SliderThumbMac):
(WebCore::SliderThumbMac::updateCellStates):
(WebCore::SliderThumbMac::rectForBounds const):
(WebCore::SliderThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlStyle.h.
(WebCore::SliderTrackMac::owningSliderTrackPart const):
* Source/WebCore/platform/graphics/mac/controls/SliderTrackMac.mm: Added.
(WebCore::SliderTrackMac::SliderTrackMac):
(WebCore::SliderTrackMac::rectForBounds const):
(WebCore::trackGradientInterpolate):
(WebCore::SliderTrackMac::draw):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::createSliderTrackPartForRenderer):
(WebCore::RenderTheme::createControlPart const):
(WebCore::RenderTheme::extractControlStyleStatesForRenderer const):
(WebCore::RenderTheme::extractControlStyleForRenderer const):
* Source/WebCore/rendering/RenderThemeMac.h:
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::TrackGradientInterpolate): Deleted.
(WebCore::RenderThemeMac::paintSliderTrack): Deleted.
(WebCore::RenderThemeMac::paintSliderThumb): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::encode):
(IPC::ArgumentCoder<ControlPart>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/258522@main
